### PR TITLE
[efr32] add support for OQPSK modulation in 915MHz band

### PR DESCRIPTION
**Pull request by zhanglongxia**
This feature enables the board BRD4170A of platform efr32 to support OQPSK modulation in 915MHz band.  The parameters of this new mode follow the definitions of IEEE802.15.4-2006 section 6.1 and section 6.8.

### CODE REVIEW

**Comment by zai107** [comment]
zai

File: examples/platforms/efr32/rail_config.c [hunk 1]
```diff
@@ -0,0 +1,184 @@
+// Copyright 2018 Silicon Laboratories, Inc.
+//
```
Comment (bukepo) [inline examples/platforms/efr32/rail_config.c]: Not sure if the license is ok to be in this folder.

Comment (zhanglongxia) [reply examples/platforms/efr32/rail_config.c]: I'm not sure too. This file is auto-generated by Hardware Configurator in Simplicity Studio.  @jwhui Can you provide some suggestions.

Comment (jwhui) [reply examples/platforms/efr32/rail_config.c]: Files that do not carry the OpenThread license header must be placed in `third_party`.

Comment (zhanglongxia) [reply examples/platforms/efr32/rail_config.c]: Updated.

File: examples/platforms/efr32/brd4170a/hal-config.h [hunk 2]
```diff
@@ -196,7 +196,6 @@
 
 // $[SERIAL]
 #define HAL_SERIAL_USART0_ENABLE              (0)
-#define BSP_SERIAL_APP_PORT                   (HAL_SERIAL_PORT_USART0)
```
Comment (jwhui) [inline examples/platforms/efr32/brd4170a/hal-config.h]: This seems unrelated to this PR - is this change intentional?

File: examples/platforms/efr32/platform-band.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+typedef struct BandConfig
+{
+    RAIL_Handle_t               mRailHandle;
+    RAIL_Config_t               mRailConfig;
+    RAILSched_Config_t          mRailSchedState;
+    const RAIL_ChannelConfig_t *mChannelConfig;
+    uint8_t                     mRailTxFifo[OT_RADIO_FRAME_MAX_SIZE + 1];
+    uint8_t                     mChannelMin;
+    uint8_t                     mChannelMax;
+} BandConfig;
```
Comment (jwhui) [inline examples/platforms/efr32/platform-band.h]: nit: prefix with `efr32` (e.g. `efr32BandConfig`)?

File: examples/platforms/efr32/radio.c [hunk 4]
```diff
@@ -174,35 +204,110 @@ void efr32RadioInit(void)
...
+void efr32BandConfigInit(void (*aEventCallback)(RAIL_Handle_t railHandle, RAIL_Events_t events))
+{
+    uint8_t index = 0;
+
+#if RADIO_SUPPORT_2P4GHZ_OQPSK
+    sBandConfigs[index].mRailConfig.eventsCallback = aEventCallback;
+    sBandConfigs[index].mRailConfig.protocol       = NULL;
+    sBandConfigs[index].mRailConfig.scheduler      = &sBandConfigs[index].mRailSchedState;
+    sBandConfigs[index].mChannelConfig             = NULL;
+    sBandConfigs[index].mChannelMin                = 11;
```
Comment (jwhui) [inline examples/platforms/efr32/radio.c]: Use `ifdefs` or `enums` rather than magic numbers?

File: examples/platforms/efr32/radio.c [hunk 5]
```diff
@@ -328,20 +456,31 @@ otError otPlatRadioReceive(otInstance *aInstance, uint8_t aChannel)
 
 otError otPlatRadioTransmit(otInstance *aInstance, otRadioFrame *aFrame)
 {
-    otError           error      = OT_ERROR_NONE;
-    RAIL_CsmaConfig_t csmaConfig = RAIL_CSMA_CONFIG_802_15_4_2003_2p4_GHz_OQPSK_CSMA;
-    RAIL_TxOptions_t  txOptions  = RAIL_TX_OPTIONS_NONE;
-    RAIL_Status_t     status;
+    otError              error         = OT_ERROR_NONE;
+    RAIL_CsmaConfig_t    csmaConfig    = RAIL_CSMA_CONFIG_802_15_4_2003_2p4_GHz_OQPSK_CSMA;
```
Comment (jwhui) [inline examples/platforms/efr32/radio.c]: Does this also apply to sub-GHz?

File: examples/platforms/efr32/radio.c [hunk 6]
```diff
@@ -373,9 +514,23 @@ otRadioFrame *otPlatRadioGetTransmitBuffer(otInstance *aInstance)
...
 {
+    RAIL_Status_t        status;
+    RAIL_SchedulerInfo_t schedulerInfo = {.priority = EFR32_SCHEDULER_SAMPLE_RSSI_PRIORITY};
     OT_UNUSED_VARIABLE(aInstance);
 
-    return (int8_t)(RAIL_GetAverageRssi(sRailHandle) >> 2);
+    RAIL_Idle(sRxBandConfig->mRailHandle, RAIL_IDLE, true);
+
+    sSampleRssiDone = true;
+    status = RAIL_StartAverageRssi(sRxBandConfig->mRailHandle, sReceiveFrame.mChannel, EFR32_RSSI_AVERAGING_TIME,
```
Comment (jwhui) [inline examples/platforms/efr32/radio.c]: How is this different than `RAIL_GetAverageRssi()`?

File: examples/platforms/efr32/radio.c [hunk 7]
```diff
@@ -771,6 +939,14 @@ static void RAILCb_Generic(RAIL_Handle_t aRailHandle, RAIL_Events_t aEvents)
         status = RAIL_Calibrate(aRailHandle, NULL, RAIL_CAL_ALL_PENDING);
         assert(status == RAIL_STATUS_NO_ERROR);
     }
+
+    if (aEvents & RAIL_EVENT_RSSI_AVERAGE_DONE)
+    {
+        sSampleRssiDone = false;
```
Comment (jwhui) [inline examples/platforms/efr32/radio.c]: Is it possible for `sSampleRssiDone` to get stuck as `true`? For example, is there anything that can abort this process?

Comment (zhanglongxia) [reply examples/platforms/efr32/radio.c]: I just checked the `RAIL` API doc, it doesn't say whether the chip will generate an interrupt when the RSSI averaging is interrupted. So it has the possibility to get stuck. To prevent the code from getting stuck, I will add a timeout mechanism.

**Comment by kfirel** [comment]
Hi
how did you generate the rail_config.c file ?
(i know  that you used the simplicity stdio but what are the configuration?) 
i want to genera same config for the EFR32FG? 

BR

kfir

**Comment by zhanglongxia** [comment]
@kfir I cut a few pictures to show how to configure EFR32FG. I hope these pictures could help you.

[rail_config.zip](https://github.com/openthread/openthread/files/3489830/rail_config.zip)

**Comment by NikhilSarnaik** [comment]
Hi @zhanglongxia

Could you share the .isc file for the simplicity studio? The pictures in rail_config.zip are missing settings for Syncword, Header and CRC

Best Regards
Nikhil

**Comment by zhanglongxia** [comment]
Hi Nikhil,
 
  The .isc file is attached.

[simple_trx_with_sw_ack_BRD4170A_OQPSK.isc.zip](https://github.com/openthread/openthread/files/3515752/simple_trx_with_sw_ack_BRD4170A_OQPSK.isc.zip)

Thanks
Zhanglong

**Comment by NikhilSarnaik** [comment]
Thanks Zhanglong.

Best Regards,
Nikhil
